File: tests/leetcode/x/ruby/0185.rb
toks = STDIN.read.split
exit if toks.empty?
idx = 0
t = toks[idx].to_i
idx += 1
cases = []
t.times do
  d = toks[idx].to_i
  e = toks[idx + 1].to_i
  idx += 2
  dept_name = {}
  d.times do
    dept_name[toks[idx].to_i] = toks[idx + 1]
    idx += 2
  end
  groups = Hash.new { |h, k| h[k] = [] }
  e.times do
    idx += 1
    name = toks[idx]
    salary = toks[idx + 1].to_i
    dept_id = toks[idx + 2].to_i
    idx += 3
    groups[dept_id] << [name, salary]
  end
  rows = []
  groups.each do |dept_id, items|
    keep = items.map { |name, salary| salary }.uniq.sort.reverse.take(3)
    items.each { |name, salary| rows << [dept_name[dept_id], -salary, name, salary] if keep.include?(salary) }
  end
  rows.sort!
  lines = [rows.length.to_s]
  rows.each { |dept, neg, name, salary| lines << "#{dept},#{name},#{salary}" }
  cases << lines.join("\n")
end
print cases.join("\n\n")

File: tests/leetcode/x/python/0185.py
import sys
from collections import defaultdict


def solve_case(departments, employees):
    dept_name = {dept_id: name for dept_id, name in departments}
    groups = defaultdict(list)
    for _, name, salary, dept_id in employees:
        groups[dept_id].append((name, salary))
    out = []
    for dept_id, items in groups.items():
        keep = set(sorted({salary for _, salary in items}, reverse=True)[:3])
        for name, salary in items:
            if salary in keep:
                out.append((dept_name[dept_id], -salary, name, salary))
    out.sort()
    rows = [str(len(out))]
    rows.extend(f"{dept},{name},{salary}" for dept, _, name, salary in out)
    return "\n".join(rows)


data = sys.stdin.read().split()
if data:
    idx = 0
    t = int(data[idx])
    idx += 1
    out = []
    for _ in range(t):
        d = int(data[idx])
        e = int(data[idx + 1])
        idx += 2
        departments = []
        for _ in range(d):
            departments.append((int(data[idx]), data[idx + 1]))
            idx += 2
        employees = []
        for _ in range(e):
            employees.append((int(data[idx]), data[idx + 1], int(data[idx + 2]), int(data[idx + 3])))
            idx += 4
        out.append(solve_case(departments, employees))
    sys.stdout.write("\n\n".join(out))
